mosaic/mosaic.go

```diff
@@ -0,0 +1,423 @@
+		{Char: '▖', Coverage: [4]bool{false, false, true, false}, CoverageMap: "  \n█ "}, // Quadrant lower left.
+		{Char: '▗', Coverage: [4]bool{false, false, false, true}, CoverageMap: "  \n █"}, // Quadrant lower right.
+		{Char: '▌', Coverage: [4]bool{true, false, true, false}, CoverageMap: "█ \n█ "},  // Left half block.
+		{Char: '▐', Coverage: [4]bool{false, true, false, true}, CoverageMap: " █\n █"},  // Right half block.
+		{Char: '▀', Coverage: [4]bool{true, true, false, false}, CoverageMap: "██\n  "},  // Upper half block (already added).
+		{Char: '▄', Coverage: [4]bool{false, false, true, true}, CoverageMap: "  \n██"},  // Lower half block (already added).
+	}
+	ComplexBlocks = []Block{
+		{Char: '▙', Coverage: [4]bool{true, false, true, true}, CoverageMap: "█ \n██"},  // Quadrant upper left and lower half.
+		{Char: '▟', Coverage: [4]bool{false, true, true, true}, CoverageMap: " █\n██"},  // Quadrant upper right and lower half.
+		{Char: '▛', Coverage: [4]bool{true, true, true, false}, CoverageMap: "██\n█ "},  // Quadrant upper half and lower left.
+		{Char: '▜', Coverage: [4]bool{true, true, false, true}, CoverageMap: "██\n █"},  // Quadrant upper half and lower right.
+		{Char: '▚', Coverage: [4]bool{true, false, false, true}, CoverageMap: "█ \n █"}, // Quadrant upper left and lower right.
+		{Char: '▞', Coverage: [4]bool{false, true, true, false}, CoverageMap: " █\n█ "}, // Quadrant upper right and lower left.
+	}
+)
+
+// Block represents different Unicode block characters.
+type Block struct {
+	Char        rune
+	Coverage    [4]bool // Which parts of the block are filled (true = filled).
+	CoverageMap string  // Visual representation of coverage for debugging.
+}
+
+// Symbol represents the symbol type to use when rendering the image.
+type Symbol string
+
+// Symbol types.
+const (
+	AllSymbols     Symbol = "all"
+	HalfSymbols    Symbol = "half"
+	QuarterSymbols Symbol = "quarter"
+)
+
+// Scale represents scale mode that will be used when rendering the image.
+type Scale uint8
+
+// Symbol types.
+const (
+	None Scale = iota
+	Fit
+	Stretch
+	Center
+)
+
+// Render mosaic with default values.
+func Render(img image.Image) string {
+	m := New()
+	return m.Render(img)
+}
+
+// Mosaic is a renderer and its options.
+type Mosaic struct {
+	outputWidth    int    // Output width.
+	outputHeight   int    // Output height (0 for auto).
+	thresholdLevel uint8  // Threshold for considering a pixel as set (0-255).
+	dither         bool   // Enable Dithering (false as default).
+	useFgBgOnly    bool   // Use only foreground/background colors (no block symbols).
+	invertColors   bool   // Invert colors.
+	scale          int    // Scale level
+	symbols        Symbol // Which symbols to use: "half", "quarter", "all".
+}
+
+// New creates and returns a [Renderer].
+func New() Mosaic {
+	return Mosaic{
+		outputWidth:    80,     // Default width.
```

Comment:
Why `80` is the default width? Shouldn't it be the image width?

Reply:
outputWidth is actually expected `columns*2` (so 40 as default). I've forgot about this property, no much intuitive lemme refactor it

Reply:
alright, updated